File: 206. Reverse Linked List/206. Reverse Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
    if (head === null) {
        return null;
    }
    let res;
    function traverse(node, prev) {
        if (node.next === null) {
            node.next = prev;
            res = node;
        }
        else {
            traverse(node.next, node);
            node.next = prev;
        }
    }
    traverse(head, null);
    return res;
};

File: 206. Reverse Linked List/206. Reverse Linked List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

void traverse(struct ListNode* node, struct ListNode* prev, struct ListNode **res) {
    if (node != NULL) {
        traverse(node -> next, node, res);
        node -> next = prev;
        return;
    }
    *res = prev;
}

struct ListNode* reverseList(struct ListNode* head) {
    struct ListNode *res;
    traverse(head, NULL, &res);
    return res;
}
